evm/src/poseidon/poseidon_stark.rs

```diff
@@ -0,0 +1,597 @@
+use std::marker::PhantomData;
+
+use plonky2::field::extension::{Extendable, FieldExtension};
+use plonky2::field::packed::PackedField;
+use plonky2::field::polynomial::PolynomialValues;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::hash::poseidon::Poseidon;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::timed;
+use plonky2::util::timing::TimingTree;
+
+use super::columns::{
+    col_input_limb, col_output_limb, full_sbox_0, full_sbox_1, partial_sbox, reg_cubed_full,
+    reg_cubed_partial, reg_input_limb, reg_output_limb, reg_power_6_full, reg_power_6_partial,
+    FILTER, HALF_N_FULL_ROUNDS, NUM_COLUMNS, N_PARTIAL_ROUNDS, POSEIDON_SPONGE_WIDTH,
+};
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cross_table_lookup::Column;
+use crate::evaluation_frame::{StarkEvaluationFrame, StarkFrame};
+use crate::stark::Stark;
+use crate::util::trace_rows_to_poly_values;
+
+pub fn ctl_data<F: Field>() -> Vec<Column<F>> {
+    let mut res: Vec<_> = (0..POSEIDON_SPONGE_WIDTH).map(col_input_limb).collect();
+    res.extend((0..POSEIDON_SPONGE_WIDTH).map(col_output_limb));
+    res
+}
+
+pub fn ctl_filter<F: Field>() -> Column<F> {
+    Column::single(FILTER)
+}
+#[derive(Copy, Clone, Default)]
+pub struct PoseidonStark<F, const D: usize> {
+    pub(crate) f: PhantomData<F>,
+}
+
+/// Information about a Poseidon operation needed for witness generation.
+impl<F: RichField + Extendable<D>, const D: usize> PoseidonStark<F, D> {
+    /// Generate the rows of the trace. Note that this does not generate the permuted columns used
+    /// in our lookup arguments, as those are computed after transposing to column-wise form.
+    fn generate_trace_rows(
+        &self,
+        inputs: Vec<[u64; POSEIDON_SPONGE_WIDTH]>,
+        min_rows: usize,
+    ) -> Vec<[F; NUM_COLUMNS]> {
+        let num_rows = (inputs.len()).max(min_rows).next_power_of_two();
+
+        let mut rows = Vec::with_capacity(num_rows);
+        for input in inputs.iter() {
+            let inps_for_perm = input
+                .iter()
+                .map(|&elt| F::from_canonical_u64(elt))
+                .collect::<Vec<F>>()
+                .try_into()
+                .unwrap();
```

Comment:
This is erroneous. As we are currently storing the first 64 bits of the U256 read at a given address, we have no guarantee that this in canonical form, and may lead to possible issues.
Though it depends on how data to be hashed may be represented, so pending further discussion.